language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java

```diff
@@ -650,24 +652,60 @@ public AnalysisResultsWrapper analyze(URI uri, String content, IssueListener iss
         userSettings.analyzerProperties);
       logger.debug("Analysis triggered on " + uri + " with configuration: \n" + configuration.toString());
 
+      IssueCollector collector = new IssueCollector();
+      ServerInfo serverInfo = serverInfoCache.get(binding.serverId);
+
       long start = System.currentTimeMillis();
       AnalysisResults analysisResults;
       try {
-        analysisResults = engine.analyze(configuration, issueListener, logOutput, null);
+        analysisResults = analyze(configuration, collector);
       } catch (GlobalUpdateRequiredException e) {
-        ServerInfo serverInfo = serverInfoCache.get(binding.serverId);
         updateServerStorage(engine, serverInfo);
         updateModuleStorage(engine, serverInfo);
-        analysisResults = engine.analyze(configuration, issueListener, logOutput, null);
+        analysisResults = analyze(configuration, collector);
       } catch (StorageException e) {
-        ServerInfo serverInfo = serverInfoCache.get(binding.serverId);
         updateModuleStorage(engine, serverInfo);
-        analysisResults = engine.analyze(configuration, issueListener, logOutput, null);
+        analysisResults = analyze(configuration, collector);
       }
+
+      new ServerIssueTracker(engine, getServerConfiguration(serverInfo), projectKey, baseDir, new ServerIssueTrackingLogger())
+        .matchAndTrack(collector.issues)
+        .forEach(issueListener::handle);
+
       int analysisTime = (int) (System.currentTimeMillis() - start);
 
       return new AnalysisResultsWrapper(analysisResults, analysisTime);
     }
+
+    private AnalysisResults analyze(ConnectedAnalysisConfiguration configuration, IssueCollector collector) {
+      return engine.analyze(configuration, collector, logOutput, null);
+    }
+  }
+
+  static class IssueCollector implements IssueListener {
+    final List<Issue> issues = new LinkedList<>();
+
+    @Override
+    public void handle(Issue issue) {
+      issues.add(issue);
+    }
+  }
+
+  private class ServerIssueTrackingLogger implements org.sonarsource.sonarlint.core.tracking.Logger {
```

Comment:
It doesn't need to be a inner class and I would rename it to something such as `IssueTrackingLoggerAdapter`.